libraries/AC_PID/AC_P_2D.h

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+/// @file   AC_P_2D.h
+/// @brief  Generic PID algorithm, with EEPROM-backed storage of constants.
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Param/AP_Param.h>
+
+/// @class  AC_P_2D
+/// @brief  Copter PID control class
+class AC_P_2D {
+public:
+
+    // Constructor for PID
+    AC_P_2D(float initial_p, float dt);
+
+    // set time step in seconds
+    void set_dt(float dt) { _dt = dt; }
+
+    //  update_all - set target and measured inputs to PID controller and calculate outputs
+    //  target and error are filtered
+    //  the derivative is then calculated and filtered
+    //  the integral is then updated based on the setting of the limit flag
+    Vector2f update_all(float &target_x, float &target_y, Vector2f measurement, float error_max, float D2_max) WARN_IF_UNUSED;
+    Vector2f update_all(float &target_x, float &target_y, Vector3f measurement, float error_max, float D2_max) WARN_IF_UNUSED {
+        return update_all(target_x, target_y, Vector2f(measurement.x, measurement.y), error_max, D2_max);
+    }
```

Comment:
as mentioned in another similar context I think I'll leave it as two separate fields in order to avoid the likely issue with callers accidentally providing a copy of the targets which would lead to the subtle problem of the targets not being limited.